modules/mcagar/docs/bdd.md

```diff
@@ -0,0 +1,224 @@
+  - Provides a way to setup an editor using a custom element. A callback function is required which should setup the element to use and then return a function to teardown the element when the tests have completed.
+- `bddSetupInShadowRoot(settings: RawEditorSettings, setupModules?: Array<() => void>, focusOnInit?: boolean)`
+  - Sets up the editor within a shadow root using the settings provided. An additional `shadowRoot` lookup function is provided in the hook return object.
+
+### Example: Using TinyHooks to test a custom plugin
+
+```js
+import { TinyHooks } from '@ephox/mcagar';
+import { Editor } from 'tinymce';
+
+import { Plugin } from '../../main/ts/Plugin';
+
+describe('My Custom Plugin', () => {
+  const hook = TinyHooks.bddSetup<Editor>({
+    plugins: 'customplugin',
+    toolbar: 'custom',
+    customplugin_setting: true
+  }, [ Plugin ]);
+
+  it('should include a custom toolbar button', () => {
+    const editor = hook.editor();
+    // Assert the editor has a custom toolbat button rendered
+    ...
+  });
+});
+```
+
+## TinyContentActions
+
+The `TinyContentActions` module provides helper functions to perform actions in the content of the editor.
+
+### Keyboard actions
+
+The following functions are available to simulate keyboard events in the editor content:
+
+- `keydown(editor: Editor, keyCode: number, modifiers?: Record<string, boolean>)`
+  - Simulates a `keydown` keyboard event against the focused element in the content of the editor.
+- `keyup(editor: Editor, keyCode: number, modifiers?: Record<string, boolean>)`
+  - Simulates a `keuyp` keyboard event against the focused element in the content of the editor.
+- `keypress(editor: Editor, keyCode: number, modifiers?: Record<string, boolean>)`
+  - Simulates a `keypress` keyboard event against the focused element in the content of the editor.
+- `keystroke(editor: Editor, keyCode: number, modifiers?: Record<string, boolean>)`
+  - Simulates `keydown`, `keypress` and then a `keyup` keyboard events against the focused element in the content of the editor.
+- `type(editor: Editor, content: string)`
+  - Simulates typing the specified `content` one character at a time in the content of the editor.
+
+## TinyUiActions
+
+The `TinyUiActions` module provides helper functions to perform actions against the User Interface (UI) of the editor.
+
+###  Dialog actions
+
+The following functions are available to open or close dialogs:
+
+- `submitDialog(editor: Editor, dialogSelector?: string))`
+  - Clicks the `Save` button in the dialog to submit changes.
+- `cancelDialog(editor: Editor, dialogSelector?: string)`
+  - Clicks the `Cancel` button in the dialog to discard changes.
+- `closeDialog(editor: Editor, dialogSelector?: string)`
+  - Clicks the `X` button in the dialog to discard changes.
+- `pWaitForDialog(editor: Editor, dialogSelector?: string)`
+  - Returns a `Promise` that will resolve when a dialog has finished opening.
+
+### Keyboard actions
```

Comment:
This appears to be duplicated?

Reply:
As discussed on slack, it's not duplicated. One is for `TinyContentActions` and one is for `TinyUiActions`. If we were using something other than markdown we could reuse parts of it, but unfortunately that can't be done.